### PR TITLE
Remove all panic due to corrupt enum

### LINKED ISSUE #19229: Clean up PANIC_DUE_TO_CORRUPT_ENUM macros
https://github.com/envoyproxy/envoy/pull/19204 replaced NOT_REACHED_GCOVR_EXCL_LINE macros with more clear PANIC_DUE_TO_CORRUPT_ENUM macros.

While more clear, the panics are still suboptimal becasue doing break; instead of return; results in production crashes.  These should be cleaned up and replaced with sensible defaults on a case by case basis.

### PR DESCRIPTION
**Pull request by tommyp1ckles**
Fixes #19229: Removes/replaces all invocations of PANIC_DUE_TO_CORRUPT_ENUM.

### CODE REVIEW

**Comment by alyssawilk** [comment]
Looks like CI still needs a look?
Throwing to Kevin for first pass.  Kevin please assign me for second once you've LGTMd?

**Comment by tommyp1ckles** [comment]
@alyssawilk Yeah just fixing coverage issues atm.

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc [hunk 1]
```diff
@@ -206,7 +206,7 @@ RouteMatcher::RouteMatcher(
       route_entry = std::make_shared<ServiceNameRouteEntryImpl>(route);
       break;
     case RouteMatch::MatchSpecifierCase::MATCH_SPECIFIER_NOT_SET:
-      PANIC_DUE_TO_CORRUPT_ENUM;
+      throw EnvoyException("RouteMatcher route specifier not set");
```
Comment (KBaichoo) [inline source/extensions/filters/network/thrift_proxy/router/router_impl.cc:209]: should we handle the corrupted enum e.g. if route_entry is still nullptr after this switch

**Review by KBaichoo — COMMENTED** [review]
here's a first pass. thanks for getting rid of these!

File: source/common/tracing/custom_tag_impl.cc [hunk 2]
```diff
@@ -148,7 +148,8 @@ CustomTagUtility::createCustomTag(const envoy::type::tracing::v3::CustomTag& tag
   case envoy::type::tracing::v3::CustomTag::TypeCase::TYPE_NOT_SET:
     break; // Panic below.
```
[the remark this answers is not in the document]
Comment (tommyp1ckles) [reply source/common/tracing/custom_tag_impl.cc:149]: Yeah makes sense

File: source/extensions/common/tap/extension_config_base.cc [hunk 3]
```diff
@@ -39,7 +39,8 @@ ExtensionConfigBase::ExtensionConfigBase(
   }
   case envoy::extensions::common::tap::v3::CommonExtensionConfig::ConfigTypeCase::
       CONFIG_TYPE_NOT_SET:
-    PANIC_DUE_TO_CORRUPT_ENUM;
+    throw EnvoyException(
+        absl::StrCat("Unexpected extenison config type case", proto_config_.config_type_case()));
```
[the remark this answers is not in the document]
Comment (tommyp1ckles) [reply source/extensions/common/tap/extension_config_base.cc:43]: Yeah I'll cover that case separately.

File: source/extensions/filters/udp/udp_proxy/hash_policy_impl.cc [hunk 4]
```diff
@@ -37,24 +39,28 @@ class KeyHashMethod : public HashPolicyImpl::HashMethod {
...
+
+HashPolicyImplPtr HashPolicyImplFactory::create(
+    const Protobuf::RepeatedPtrField<
+        envoy::extensions::filters::udp::udp_proxy::v3::UdpProxyConfig_HashPolicy>& hash_policies) {
   ASSERT(hash_policies.size() == 1);
-  switch (hash_policies[0]->policy_specifier_case()) {
+  switch (hash_policies[0].policy_specifier_case()) {
   case UdpProxyConfig::HashPolicy::PolicySpecifierCase::kSourceIp:
-    hash_impl_ = std::make_unique<SourceIpHashMethod>();
+    return std::make_unique<HashPolicyImpl>(HashPolicyImpl(std::make_unique<SourceIpHashMethod>()));
```
Comment (KBaichoo) [inline source/extensions/filters/udp/udp_proxy/hash_policy_impl.cc:53]: a bit confused why you're  creating the HashPolicyImpl twice? once via the `HashPolicyImpl(HashMethodPtr hash_impl)` ctor and another via the unique_ptr? Here and below

File: source/extensions/filters/udp/udp_proxy/hash_policy_impl.h [hunk 5]
```diff
@@ -32,10 +27,27 @@ class HashPolicyImpl : public Udp::HashPolicy {
 
   using HashMethodPtr = std::unique_ptr<HashMethod>;
 
+  HashPolicyImpl(HashMethodPtr hash_impl) : hash_impl_(std::move(hash_impl)) {}
```
Comment (KBaichoo) [inline source/extensions/filters/udp/udp_proxy/hash_policy_impl.h:30]: is this meant to be private to enforce creation through the factory?

File: source/extensions/common/matcher/matcher.cc [hunk 6]
```diff
@@ -149,7 +153,10 @@ HttpGenericBodyMatcher::HttpGenericBodyMatcher(
       patterns_->push_back(i.string_match());
       break;
     case envoy::config::common::matcher::v3::HttpGenericBodyMatch::GenericTextMatch::RULE_NOT_SET:
-      PANIC_DUE_TO_CORRUPT_ENUM;
+      throw EnvoyException("http generic body match rule not set");
+    default:
+      throw EnvoyException(
+          absl::StrCat("unexpected http generic body match rule type", i.rule_case()));
```
Comment (KBaichoo) [inline source/extensions/common/matcher/matcher.cc:159]: you probably want a ": " after type

File: source/extensions/common/tap/extension_config_base.cc [hunk 7]
```diff
@@ -16,13 +16,14 @@ ExtensionConfigBase::ExtensionConfigBase(
     : proto_config_(proto_config), config_factory_(std::move(config_factory)), tls_slot_(tls) {
   tls_slot_.set([](Event::Dispatcher&) { return std::make_shared<TlsFilterConfig>(); });
 
+  // switch (proto_config_.config_type_case()) {
```
Comment (KBaichoo) [inline source/extensions/common/tap/extension_config_base.cc:19]: rm

File: source/extensions/filters/http/ip_tagging/ip_tagging_filter.h [hunk 8]
```diff
@@ -46,23 +46,28 @@ class IpTaggingFilterConfig {
   void incTotal() { incCounter(total_); }
 
 private:
-  static FilterRequestType requestTypeEnum(
+  /**
+   * Gets request type enum from IP tagging proto.
+   * NOTE: Does not emit ENVOY_ERROR when unexpected proto request_type is passed.
```
Comment (KBaichoo) [inline source/extensions/filters/http/ip_tagging/ip_tagging_filter.h:51]: s/ENVOY_ERROR/ENVOY_BUG/ ?

File: source/common/upstream/load_balancer_impl.h [hunk 9]
```diff
@@ -305,7 +305,7 @@ class ZoneAwareLoadBalancerBase : public LoadBalancerBase {
   /**
    * Index into priority_set via hosts source descriptor.
    */
-  const HostVector& hostSourceToHosts(HostsSource hosts_source) const;
+  const HostVector* hostSourceToHosts(HostsSource hosts_source) const;
```
Comment (KBaichoo) [inline source/common/upstream/load_balancer_impl.h:308]: worth documenting this could return nullptr?

**Review by KBaichoo — COMMENTED** [review]
Here's another pass

File: source/extensions/filters/udp/udp_proxy/hash_policy_impl.h (hunk 5 above)
Comment (tommyp1ckles) [reply source/extensions/filters/udp/udp_proxy/hash_policy_impl.h:30]: The factory is a different class so I think I need a public constructor.

Comment (tommyp1ckles) [reply source/extensions/filters/udp/udp_proxy/hash_policy_impl.h:30]: I might try to embed the factory class.

**Review by KBaichoo — COMMENTED** [review]
Few small comments. Thanks this is probably ready for new eyes after this

**Comment by KBaichoo** [comment]
/assign @alyssawilk

for another pass

File: source/common/common/assert.h [hunk 10]
```diff
@@ -261,10 +261,5 @@ void resetEnvoyBugCountersForTest();
 
 #define PANIC_DUE_TO_PROTO_UNSET PANIC("unset oneof")
 
-// Envoy has a number of switch statements which panic if there's no legal value set.
-// This is not encouraged, as it's too easy to panic using break; instead of return;
-// but this macro replaces a less clear crash using NOT_REACHED_GCOVR_EXCL_LINE.
-#define PANIC_DUE_TO_CORRUPT_ENUM PANIC("corrupted enum");
```
Comment (alyssawilk) [inline source/common/common/assert.h:267]: WOW!  So awesome to have all of these cleaned up!

File: source/common/upstream/health_checker_impl.cc [hunk 11]
```diff
@@ -450,7 +450,7 @@ HttpHealthCheckerImpl::codecClientType(const envoy::type::v3::CodecClientType& t
   case envoy::type::v3::HTTP1:
     return Http::CodecType::HTTP1;
   }
-  PANIC_DUE_TO_CORRUPT_ENUM
+  throw EnvoyException("unexpected code client type");
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/common/upstream/health_checker_impl.cc:453]: ping?
though I think this is called on the data plane so ENVOY_BUG-and-default-to-http1 may be the way to go?

File: source/extensions/filters/network/thrift_proxy/conn_manager.cc [hunk 12]
```diff
@@ -656,13 +656,15 @@ void ConnectionManager::ActiveRpc::prepareFilterAction(DecoderEvent event, absl:
...
     };
-    break;
+    return;
   case DecoderEvent::SetEnd:
     filter_action_ = [](DecoderEventHandler* filter) -> FilterStatus { return filter->setEnd(); };
-    break;
-  default:
-    PANIC_DUE_TO_CORRUPT_ENUM;
+    return;
+  case DecoderEvent::ContinueDecode:
```
Comment (alyssawilk) [inline source/extensions/filters/network/thrift_proxy/conn_manager.cc:663]: cc @zuercher @rgs1

File: source/extensions/stat_sinks/common/statsd/statsd.cc [hunk 13]
```diff
@@ -62,15 +62,23 @@ void UdpStatsdSink::flush(Stats::MetricSnapshot& snapshot) {
 
   for (const auto& counter : snapshot.counters()) {
```
Comment (alyssawilk) [inline source/extensions/stat_sinks/common/statsd/statsd.cc:63]: note to self: stopped here

**Review by alyssawilk — COMMENTED** [review]
Whew!  Got about 3/4 of the way through but I'll have to pick up the rest tomorrow. This is SUCH awesome clean up, but next time let's definitely split this into a few PRs :-)

**Comment by tommyp1ckles** [comment]
Haven't had time to work in this recently. Will try to move it forward next week.

**Pull request closed without merging**